File: LeetCode/Python Solutions/LRU Cache/cache.py
from collections import OrderedDict

class LRUCache:

    def __init__(self, capacity: int):
        # Initialize the OrderedDict and set the capacity
        self.cache = OrderedDict()
        self.capacity = capacity

    def get(self, key: int) -> int:
        # If the key is not in the cache, return -1
        if key not in self.cache:
            return -1
        
        # Move the key-value pair to the end of the cache to mark it as recently used
        self.cache.move_to_end(key)
        
        # Return the value associated with the key
        return self.cache[key]

    def put(self, key: int, value: int) -> None:
        # If the key is already in the cache, move it to the end to mark it as recently used
        if key in self.cache:
            self.cache.move_to_end(key)

        # Insert the key-value pair into the cache
        self.cache[key] = value

        # If the cache has exceeded its capacity, remove the least recently used key-value pair
        if len(self.cache) > self.capacity:
            self.cache.popitem(last=False)

if __name__ == '__main__':
    lru_cache = LRUCache(2)
    lru_cache.put(1, 1)
    lru_cache.put(2, 2)
    assert lru_cache.get(1) == 1
    lru_cache.put(3, 3)
    assert lru_cache.get(2) == -1
    lru_cache.put(4, 4)
    assert lru_cache.get(1) == -1
    assert lru_cache.get(3) == 3
    assert lru_cache.get(4) == 4

File: LeetCode/Java Solutions/LRU Cache/cache.java
import java.util.LinkedHashMap;
import java.util.Map;

class LRUCache {
    private LinkedHashMap<Integer, Integer> cache;  // LinkedHashMap used to store the key-value pairs
    private int capacity;  // Maximum capacity of the cache

    public LRUCache(int capacity) {
        // Initialize the cache with the specified capacity and a load factor of 0.75
        // The last parameter "true" indicates that the access order should be maintained
        this.cache = new LinkedHashMap<>(capacity, 0.75f, true) {
            // Override removeEldestEntry to implement LRU eviction strategy
            protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
                // Return true if the size of the cache exceeds the capacity
                return size() > capacity;
            }
        };
        this.capacity = capacity;  // Set the maximum capacity of the cache
    }

    public int get(int key) {
        // Return the value associated with the given key, or -1 if not found
        return cache.getOrDefault(key, -1);
    }

    public void put(int key, int value) {
        // Add or update the key-value pair in the cache
        cache.put(key, value);
    }
}

public class Main {
    public static void main(String[] args) {
        LRUCache lruCache = new LRUCache(2);  // Create an instance of the LRUCache with capacity 2
        lruCache.put(1, 1);  // Add key-value pair (1, 1) to the cache
        lruCache.put(2, 2);  // Add key-value pair (2, 2) to the cache
        assert lruCache.get(1) == 1;  // Retrieve the value associated with key 1, should be 1
        lruCache.put(3, 3);  // Add key-value pair (3, 3) to the cache, evicting (2, 2) as it was the least recently used
        assert lruCache.get(2) == -1;  // Key 2 is not found in the cache, should return -1
        lruCache.put(4, 4);  // Add key-value pair (4, 4) to the cache, evicting (1, 1) as it was the least recently used
        assert lruCache.get(1) == -1;  // Key 1 is not found in the cache, should return -1
        assert lruCache.get(3) == 3;  // Retrieve the value associated with key 3, should be 3
        assert lruCache.get(4) == 4;  // Retrieve the value associated with key 4, should be 4
    }
}
